[FLINK-6938][FLINK-6939] [cep] Not store IterativeCondition with NFA state and support RichFunction interface

wuchong opened the pull request:
The core idea is that the `StateTransition` is unique in a NFA graph. So we store the conditions with a map which mapping from `StateTransition` to `IterativeCondition`, so the conditions can not serialized with NFA state. If I missed something, please point out.

This PR also includes FLINK-6938: IterativeCondition supports RichFunction interface. 


Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

dawidwys commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ConditionRegistry.java:
```diff
@@ -0,0 +1,135 @@
+		}
+
+		TransitionInfo(StateTransition<?> transition) {
+			this.action = transition.getAction();
```
Could we use `update()` here?

wuchong replied:
Yes, thanks.

dawidwys commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java:
```diff
@@ -126,6 +129,7 @@ public AbstractKeyedCEPPatternOperator(
 		this.isProcessingTime = Preconditions.checkNotNull(isProcessingTime);
 		this.keySerializer = Preconditions.checkNotNull(keySerializer);
 		this.nfaFactory = Preconditions.checkNotNull(nfaFactory);
+		this.conditionRegistry = nfaFactory.createNFA().getConditionRegistry();
```
Maybe let's access conditionRegistry through nfaFactory rather than NFA itselfs?

Anyway I think we should think of removing one layer in the `NFACompiler.java`, but I will create a seperate JIRA for this.

wuchong replied:
I have tried that, but found it would touch too many codes.  We need to change `process` signature to add the `conditionRegistry` parameter, this need to change many codes. And we use `NFACompiler.compile` to compile the given pattern into a  NFA, but without `conditionRegistry`, that's why I have to put `conditionRegistry` in NFA.

This may look a bit wired. But I'm not sure which way is better.

dawidwys replied:
I am ok that `NFA` right now has pointer to `conditionRegistry`. 

The thing I meant is just in this line, as we call `createNFA()` and ignore the created `NFA`. The `nfaFactory` also has a pointer to the same `nfaRegistry` so instead of `nfaFactory.createNFA().getConditionRegistry()` we could call `nfaFactory.getConditionRegistry()`. What do you think?

wuchong replied:
Yes, we can.

wuchong commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java:
```diff
@@ -1131,6 +1152,7 @@ private void serializeStates(Set<State<T>> states, DataOutputView out) throws IO
 					actionSerializer.serialize(transition.getAction(), out);
 
+					// backward compatibility
 					serializeCondition(transition.getCondition(), out);
```
I'm not sure. Do we need to backward compatible to previous version (like 1.3.0 which serializes conditions) ?

dawidwys replied:
Yes, but I think we could increase the VERSION field, and wrap this line in `if (readVersion == 1) {...}`

wuchong replied:
Thank you for the suggestion.

wuchong commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/StateTransition.java:
```diff
@@ -36,6 +36,12 @@
+	 * Now the condition is stored in {@link ConditionRegistry}.
+	 */
+	@Deprecated
 	private IterativeCondition<T> newCondition;
```
I think so. I will check it again.

dianfu commented:
Nice work. Just one comment: the states in NFA is metadata, maybe we should get rid of serializing them at all?

wuchong commented:
@dianfu I agree with you. I think we can improve that in another JIRA.
@dawidwys I have addressed all the comments.
@kl0u It'll be great if you can have a look. 

Cheers,
Jark

dawidwys commented:
@dianfu I agree that we should consider removing states from serialization, but I have some concerns about it:

1. `State`s are part of `ComputationState`s which describe partial matches, which means they are definitely not metadata. If we do not want to store `State`s, I think we will be forced to either drop checkpointing of partial matches or implement logic of translating them to newly constructed graph of states, which I think would be hard if possible at all.

2. One feature I would like to see in FlinkCEP library is dynamically providing new Patterns(e.g. through Side Inputs), I think in this case `State`s serializing will be crucial.

dawidwys commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/StateTransition.java:
```diff
@@ -36,24 +34,14 @@
-	 * Now the conditions extend the {@link IterativeCondition}.
-	 */
-	@Deprecated
-	private FilterFunction<T> condition;
```
I think this field needs to stay, though. It was prior to 1.3.x, so it is needed for java serialization, which is used for snapshots taken in 1.2.x

dawidwys commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/StateTransition.java:
```diff
@@ -98,32 +82,9 @@ public int hashCode() {
-	 * A wrapper to transform a {@link FilterFunction} into a {@link SimpleCondition}.
-	 * This is used only when migrating from an older Flink version.
-	 */
-	private static class FilterWrapper<T> extends SimpleCondition<T> {
```
Same as above.

dianfu commented:
@dawidwys In my mind, to achieve `translating them to newly constructed graph of states`, we only need to get the state from the NFA by the `state name` stored in `ComputationState`. Not sure if my understanding is correct?
For `dynamically providing new Patterns`, do you mean dynamically changing the NFA? Another approach coming into my mind is that we can store the states in another `ValueState`(compared with the partially matched events stored) and only update it if the NFA changes. What's your thought?

dawidwys commented:
@dianfu I've created a JIRA https://issues.apache.org/jira/browse/FLINK-6983 for it. Let's move the discussion there.

kl0u commented:
Hi @wuchong ! Nice work! 

I have some changes to propose in the code itself, but before moving on with this, I have some questions about the need of the change. 

For the second, I have commented on the JIRA (FLINK-6939) and it would be nice to finish the discussion there first, before checking the following comments.

Now for the comments on the code I would suggest the following:

1) Remove the `TransitionInfo` from the `ConditionRegistry` and replace it with the `StateTransition<T>`, as the `TransitionInfo` is just a copy of the state transition. Also make the
`ConditionRegistry`, `ConditionRegistry<T>`.
2)  Pass the `ConditionRegistry` to the `State` itself in the constructor. This will simplify all the changes of the `addStateTransition()`, `addProceed()`... as they do not need the registry as an additional argument anymore.
3) If we agree to keep the `RichIterativeFunction`, then we should probably think if we can remove the non-rich `IterativeCondition` (although this is optional).

For these changes, I have a branch where I tested them a bit but of course feel free to make the code nicer, and also tell me if I did any change that does not seem reasonable to you. This branch was just a fast testing branch on my side. You can find it here:

https://github.com/kl0u/flink/tree/cep-iter-pr

Again thanks a lot for the work and I am looking forward to you comments!
